app/src/parser_impl.c

```diff
@@ -0,0 +1,873 @@
+    CHECK_ERROR(_readArrayU64(c, application->foreign_assets, &application->num_foreign_assets, COUNT(application->foreign_assets)))
+    addItem(3);
+
+    CHECK_ERROR(_findKey(c, KEY_APP_ACCOUNTS))
+    CHECK_ERROR(_readAccounts(c, application->accounts, &application->num_accounts, COUNT(application->accounts)))
+    addItem(4);
+    addItem(5);
+
+    CHECK_ERROR(_findKey(c, KEY_APP_ARGS))
+    CHECK_ERROR(_readAppArgs(c, application->app_args, application->app_args_len, &application->num_app_args,
+                             COUNT(application->app_args)))
+    addItem(6);
+    addItem(7);
+
+    CHECK_ERROR(_findKey(c, KEY_APP_GLOBAL_SCHEMA))
+    CHECK_ERROR(_readStateSchema(c, &application->global_schema))
+    addItem(8);
+
+    CHECK_ERROR(_findKey(c, KEY_APP_LOCAL_SCHEMA))
+    CHECK_ERROR(_readStateSchema(c, &application->local_schema))
+    addItem(9);
+
+    CHECK_ERROR(_findKey(c, KEY_APP_APROG_LEN))
+    CHECK_ERROR(_readBin(c, application->aprog, (uint16_t*)&application->aprog_len, sizeof(application->aprog)))
+    addItem(10);
+
+    CHECK_ERROR(_findKey(c, KEY_APP_CPROG_LEN))
+    CHECK_ERROR(_readBin(c, application->cprog, (uint16_t*)&application->cprog_len, sizeof(application->cprog)))
+    addItem(11);
+
+
+    return parser_ok;
+}
+
+parser_error_t _readArray_args(parser_context_t *c, uint8_t args[][MAX_ARGLEN], size_t args_len[], uint8_t *argsSize, uint8_t maxArgs)
+{
+    uint8_t byte = 0;
+    CHECK_ERROR(_readUInt8(c, &byte))
+    if (byte < FIXARR_0 || byte > FIXARR_15) {
+        return parser_msgpack_array_unexpected_size;
+    }
+    *argsSize = byte - FIXARR_0;
+
+    if(*argsSize > maxArgs) {
+        return parser_msgpack_array_too_big;
+    }
+
+    for (size_t i = 0; i < *argsSize; i++) {
+        CHECK_ERROR(_readBin(c, args[i], (uint16_t*)&args_len[i], sizeof(args[0])))
+    }
+    return parser_ok;
+}
+
+parser_error_t _read(parser_context_t *c, parser_tx_t *v)
+{
+    size_t keyLen = 0;
+    initializeItemArray();
+
+    CHECK_ERROR(_readMapSize(c, &keyLen))
+    if(keyLen > UINT8_MAX) {
+        return parser_unexpected_number_items;
+    }
+
+    // Read Tx type
+    CHECK_ERROR(_readTxType(c, v))
+
+    // Read common params
+    CHECK_ERROR(_readTxCommonParams(c, v));
+
+    // Read Tx specifics params
+    switch (v->type) {
+    case TX_PAYMENT:
+        CHECK_ERROR(_readTxPayment(c, v))
+        break;
+    case TX_KEYREG:
+        CHECK_ERROR(_readTxKeyreg(c, v))
+        break;
+    case TX_ASSET_XFER:
+        CHECK_ERROR(_readTxAssetXfer(c, v))
+        break;
+    case TX_ASSET_FREEZE:
+        CHECK_ERROR(_readTxAssetFreeze(c, v))
+        break;
+    case TX_ASSET_CONFIG:
+        CHECK_ERROR(_readTxAssetConfig(c, v))
+        break;
+    case TX_APPLICATION:
+        CHECK_ERROR(_readTxApplication(c, &v->application))
+        break;
+    default:
+        return paser_unknown_transaction;
+        break;
+    }
+
+    num_items = common_num_items + tx_num_items + 1;
```

Comment:
in **keyLen** we have the quantity of args that are in the blob. However, some of them will not be displayed and other might take a default value (not in blob but they must be counted). Not the cleanest solution but each part (common and txn specifi) will take care of counting how many elements are present.
Adding +1 at the end is because txn_type is not considered in common params